src/Pgvector/Npgsql/VectorHandler.cs

```diff
@@ -4,47 +4,72 @@
 using Npgsql.Internal.TypeHandling;
 using Npgsql.PostgresTypes;
 using System;
+using System.Diagnostics;
 using System.Threading;
 using System.Threading.Tasks;
 
 namespace Pgvector.Npgsql
 {
-    public partial class VectorHandler : NpgsqlSimpleTypeHandler<Vector>
+    public partial class VectorHandler : NpgsqlTypeHandler<Vector>
     {
         public VectorHandler(PostgresType pgType) : base(pgType) { }
 
-        public override Vector Read(NpgsqlReadBuffer buf, int len, FieldDescription fieldDescription = null)
+        public override async ValueTask<Vector> Read(NpgsqlReadBuffer buf, int len, bool async, FieldDescription fieldDescription = null)
         {
+            await buf.Ensure(2 * sizeof(ushort), async);
             var dim = buf.ReadUInt16();
             var unused = buf.ReadUInt16();
             if (unused != 0)
                 throw new InvalidCastException("expected unused to be 0");
 
             var vec = new float[dim];
-            for (int i = 0; i < dim; i++)
+            for (var i = 0; i < dim; i++)
+            {
+                await buf.Ensure(sizeof(float), async);
```

Comment:
That would effectively make things the same as they were before, where we attempt to pre-buffer the entire vector data up-front; and if the vector is very large, this would fail as it wouldn't fit in Npgsql's internal buffer (8k by default).

buf.Ensure is generally a very light operation that should get inlined, and is very unlikely to have any sort of performance impact. In theory, if we really want to micro-optimize, we could pre-buffer by batches of a certain number, e.g. doing Ensure only once every 64 values. But I wouldn't introduce this complexity without a benchmark clearly showing that it's worth it.